include/LCObjects/LCTrack.h

```diff
@@ -0,0 +1,215 @@
+
+inline LCTrack::LCTrack(const LCTrackParameters &parameters) :
+    object_creation::Track::Object(parameters),
+    m_trackStates()
+{
+    for (auto const& inputTrackState: parameters.m_trackStates ) {
+        m_trackStates.push_back( inputTrackState.Get() );
+    }
+
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline const LCTrackStates &LCTrack::GetTrackStates() const
+{
+    return m_trackStates;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline LCTrackFactory::Parameters *LCTrackFactory::NewParameters() const
+{
+    return (new LCTrackParameters);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline pandora::StatusCode LCTrackFactory::Create(const Parameters &parameters, const Object *&pObject) const
+{
+    const LCTrackParameters &lcTrackParameters(dynamic_cast<const LCTrackParameters&>(parameters));
+    pObject = new LCTrack(lcTrackParameters);
+
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline pandora::StatusCode LCTrackFactory::Read(Parameters &parameters, pandora::FileReader &fileReader) const
+{
+    // ATTN: To receive this call-back must have already set file reader track factory to this factory
+    LCInputTrackStates trackStates;
+    int nTrackStates;
+    if (pandora::BINARY == fileReader.GetFileType())
+    {
+        pandora::BinaryFileReader &binaryFileReader(dynamic_cast<pandora::BinaryFileReader&>(fileReader));
+        PANDORA_RETURN_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, binaryFileReader.ReadVariable(nTrackStates));
+        for (int i = 0; i < nTrackStates; ++i)
+        {
+            pandora::TrackState trackState(0.0,0.0,0.0,0.0,0.0,0.0);
+            PANDORA_RETURN_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, binaryFileReader.ReadVariable(trackState));
+            trackStates.push_back( pandora::InputTrackState(trackState) );
+        }
+    }
+    else if (pandora::XML == fileReader.GetFileType())
+    {
+        pandora::XmlFileReader &xmlFileReader(dynamic_cast<pandora::XmlFileReader&>(fileReader));
+        PANDORA_RETURN_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, xmlFileReader.ReadVariable("NumberOfTrackStates", nTrackStates));
+        for (int i = 0; i < nTrackStates; ++i)
+        {
+            pandora::TrackState trackState(0.0,0.0,0.0,0.0,0.0,0.0);
+            PANDORA_RETURN_RESULT_IF(pandora::STATUS_CODE_SUCCESS, !=, xmlFileReader.ReadVariable("TrackState", trackState));
```

Comment:
I'm not sure that the xml parser will like this. It may just read the same (first) TrackState tag multiple times. I think the xml parsing might need to be handled at a lower level here (unfortunately), with the xml node being forced on to its "NextSibling()"

It would actually be quite good to have these functions in place (although it isn't mandatory), because Steve and I foresee using Pandora persistency and the Pandora standalone development environment as a preferred way of maintaining the LC reconstruction e.g. testing that there's no degradation in LC performance when we make LArTPC-motivated changes to the SDK.

Testing is a bit involved if you've not done it before. What you need to do (or Steve might have time to take a look; we haven't talked about this in detail yet) is implement a LC-specific version of the SDK-provided EventWriting and EventReading algorithms (could inherit most functionality). You'd then need to register your factory objects with the EventWriter/EventReader instance.

To test, you'd then run a LC job, with an EventWriting algorithm instance running as e.g. the first (might as well, for clarity) algorithm. You can then reproduce the same reconstruction input, then output, using the LCReco entry point application (see LCReco on GitHub), running a standard PandoraSettings file, but with an instance of the EventReading algorithm first up.

We can discuss in more detail, as required. At a quick look, I think the binary version is OK, but the xml version will probably need the change described above.

Reply:
Then instead, writing out TrackState0,TrackState1,...,TrackStateN ?

Reply:
Yes, this is a good, pragmatic way forwards.

Reply:
Can we make the implementation of the LCEventWriting/LCEventReading algorithms another PR in the near-ish future?
From what I see is that one needs, e.g., an LCEventWritingAlgorithm, which instantiates, e.g., an LCXMLFileWriter object, which re-implements the WriteTrack function[*]. 

Or is there a different way to "register the factory with the EventWriter instance"?

[*] Plus the permutations for Binary and Reader.

Reply:
We can separate out the two aspects if you prefer (i. getting access to the additional track states, and ii. sorting out the persistency).

There's some examples (a little bit involved) of how we have done this for the LArTPC use case, linked below. The idea is that you just need to register the factories with the pandora:: FileWriter and pandora::FileWriter, then proceed as before.

https://github.com/PandoraPFA/LArContent/blob/master/larpandoracontent/LArPersistency/EventWritingAlgorithm.cc#L113
https://github.com/PandoraPFA/LArContent/blob/master/larpandoracontent/LArPersistency/EventReadingAlgorithm.cc#L140

When writing or reading e.g. a Track, the existing derived binary and xml implementations will pass control to the registered factory to do their "extra" bit of writing/reading before working with the standard object properties described in the SDK.

https://github.com/PandoraPFA/PandoraSDK/blob/master/src/Persistency/BinaryFileWriter.cc#L274
https://github.com/PandoraPFA/PandoraSDK/blob/master/src/Persistency/XmlFileWriter.cc#L283

https://github.com/PandoraPFA/PandoraSDK/blob/master/src/Persistency/BinaryFileReader.cc#L650
https://github.com/PandoraPFA/PandoraSDK/blob/master/src/Persistency/XmlFileReader.cc#L617

Please just let us know how you want to proceed and when you think things are ready to be merged. We'll then take a quick look again and hopefully make the merge (maybe making a clear note to finalise the persistency at a later date, as required).

Reply:
I have created #17 (assign to me?) to keep a reminder for the LC EventWriting algorithm, which I would try to tackle this week.
Otherwise this code is ready to be merged as far as I am concerned, please have another look.
We can then run larger scale tests.